clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1193,12 +1205,18 @@ event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst
     // ifetch and not any of the expansion instrs.  We instrument the first
     // one to handle a zero-iter loop.  For offline, we just record the pc; for
     // online we have to ignore "instr" here in instru_online::instrument_instr().
-    if (!ud->repstr || drmgr_is_first_nonlabel_instr(drcontext, instr)) {
+    if (!(ud->repstr || ud->scatter_gather) ||
+        drmgr_is_first_nonlabel_instr(drcontext, instr)) {
         adjust = instrument_instr(drcontext, tag, ud, bb, instr, reg_ptr, adjust, instr);
     }
     ud->last_app_pc = instr_get_app_pc(instr);
 
-    if (is_memref) {
+    // For the scatter/gather expansion, the first instr turns out to be a non-app
+    // memref (a reg spill). So, we reach here for a non-app instruction and we
+    // don't want to add a memref for that.
+    bool is_first_instr_in_scatter_gather_expansion =
+        ud->scatter_gather && drmgr_is_first_nonlabel_instr(drcontext, instr);
```

Comment:
Should this also check it's a non-app instr?  What if something changes (the expansion or the logic here to skip non-app instrs up above) -- we won't want to skip an app instr here.  This could would ideally be changed alongside a change like that: maybe a xref to #4865.

Reply:
Done.